## Adquisición y preprocesamiento de la información

<table>
    <tr>
      <td>Grado en Ingeniería de Datos e Inteligencia Artificial - Facultad de Informática - UCM
      </td>
      <td>
      <img src="https://biblioteca.ucm.es/data/cont/media/www/pag-88746//escudo.jpg"  width=50/>
      </td>
     </tr>
</table>




### Práctica 5 -Solución

Comenzamos cargando algunos módulos que nos van a venir bien

In [1]:
pip install --user --upgrade geopy

Note: you may need to restart the kernel to use updated packages.


In [2]:
modules = ["geopy","folium","pprint"]


import sys
import os.path
from subprocess import check_call
import importlib
import os

def instala(modules):
    print("Instalando módulos")
    for m in modules:
        # para el import quitamos [...] y ==...
        p = m.find("[")
        mi = m if p==-1 else m[:p]
        p = mi.find("==")
        mi = mi if p==-1 else mi[:p]
        torch_loader = importlib.util.find_spec(mi)
        if torch_loader is not None:
            print(m," encontrado")
        else:
            print(m," No encontrado, instalando...",end="")  
            try:        
                r = check_call([sys.executable, "-m", "pip", "install", "--user", m])
                print("¡hecho!")
            except:
                print("¡Problema al instalar ",m,"! ¿seguro que el módulo existe?",sep="")

    print("¡Terminado!")

instala(modules)  

Instalando módulos
geopy  encontrado
folium  encontrado
pprint  encontrado
¡Terminado!


Para mostrar dataframes uno al lado de otro

In [3]:
from IPython.display import display, HTML
from IPython.display import display_html
from IPython.display import Markdown

def display_side_by_side(*args,title=""):
    print(title)
    html_str = ''
    for df in args:
        html_str += '&nbsp;&nbsp;&nbsp;'+df.to_html()
    display_html(
        html_str.replace('table','table style="display:inline"'), 
        raw=True)

## Funciones 

A partir de ahora vamos a definir nuestras propias funciones. Aparte del uso habitual de encapsulación y reusabilidad, en ingeniería de datos van a ser útiles para hacer operaciones complejas sobre dataframes.

En un ejemplo sencillo de función sin parámetros que devuelve directamente un valor:

In [4]:
from pathlib import Path
def get_path():
    return Path.cwd() 

### llamada a la función
path = get_path()
path    

WindowsPath('d:/docencia/2223/APD/pruebas')

Otro ejemplo un poco más interesante: la función recibe una url como parámetro y devuelve un dataframe después de cargarlo y hacer algo de preprocesamiento simple:

In [5]:
import pandas as pd

# lee el fichero con opinión de usuarios sobre los candidatos Clinton y Trump en las elecciones americanas de 2016
def lee_usa(url):
    df = pd.read_csv(url).iloc[:,1:]   # df es una variable local de la función
    df.columns = ["id","clinton","trump"]
    return df

### llamada a la función
url="https://github.com/RafaelCaballero/tdm/raw/master/datos/user10CT.zip"
df_usa = lee_usa(url)
df_usa

id   clinton     trump
0       729822791747444737 -0.567568  0.415094
1               3438745348 -0.666667 -0.111111
2               1167485504 -0.466667  0.500000
3                219793074 -0.729730  0.620370
4               4605113535 -0.586207  0.373832
...                    ...       ...       ...
161168           118676558  0.000000  0.000000
161169          1438467932  0.857143  0.000000
161170  772542459901571072 -0.583333  0.000000
161171          3833168952 -0.400000  0.000000
161172  797601605449154560  0.017241  0.000000

[161173 rows x 3 columns]

En ocasiones las funciones no devuelven nada, no tienen `return`. Cuando esto sucede es porque la función tiene "efectos laterales": por ejemplo crear una carpeta o, como en este caso, mostrar por pantalla el contenido de una carpeta

In [6]:

# lista los ficheros y subcarpetas del path
def lista(path):
    print(path)
    espacio = " "*5
    for p in path.iterdir():
        if p.is_dir():
            print(espacio,p.name,": carpeta",sep="")
        else:
            if p.is_file():
                print(espacio,p.name,": fichero",sep="")
    print("="*40)
                
#### llamada
lista(path)

d:\docencia\2223\APD\pruebas
     .cache: fichero
     .ipynb_checkpoints: carpeta
     A.csv: fichero
     B.csv: fichero
     BeautifulSoup-Copy1.ipynb: fichero
     BeautifulSoup.ipynb: fichero
     C.csv: fichero
     cancion.mp3: fichero
     datosparo.ipynb: fichero
     IMDB.ipynb: fichero
     mapa.ipynb: fichero
     meltExample.ipynb: fichero
     practica1Sol.ipynb: fichero
     practica3sol.ipynb: fichero
     raw: carpeta
     repartidos: carpeta
     requiem.mp3: fichero
     satelites.ipynb: fichero
     spotifySol.ipynb: fichero
     top_25_pandas_tricks.ipynb: fichero
     Untitled.ipynb: fichero
     Untitled1.ipynb: fichero
     Untitled2.ipynb: fichero
     Untitled3.ipynb: fichero
     usa.ipynb: fichero


**Ejercicio 1**
A partir de la solución de la pregunta 1 en la práctica 4 escribir una función que reciba 2 parámetros:

*Nombre*: `crea_carpetas(path,folders)`
- `path`: el path donde se quieren crear subcarpetas
- `folders`: una lista de subcarpetas a crear en el path.

*Salida*: no devuelve nada, solo crea las subcarpetas dentro del path si no existen

In [7]:
## Solución 
def crea_carpetas(path,folders):
    for f in folders:
        newFolder = Path(path,f)
        newFolder.mkdir(exist_ok=True,parents=True)

#### no modificar de aquí hacia abajo
# llamada 
crea_carpetas(path,["raw","repartidos"])
# para probar
lista(path)

d:\docencia\2223\APD\pruebas
     .cache: fichero
     .ipynb_checkpoints: carpeta
     A.csv: fichero
     B.csv: fichero
     BeautifulSoup-Copy1.ipynb: fichero
     BeautifulSoup.ipynb: fichero
     C.csv: fichero
     cancion.mp3: fichero
     datosparo.ipynb: fichero
     IMDB.ipynb: fichero
     mapa.ipynb: fichero
     meltExample.ipynb: fichero
     practica1Sol.ipynb: fichero
     practica3sol.ipynb: fichero
     raw: carpeta
     repartidos: carpeta
     requiem.mp3: fichero
     satelites.ipynb: fichero
     spotifySol.ipynb: fichero
     top_25_pandas_tricks.ipynb: fichero
     Untitled.ipynb: fichero
     Untitled1.ipynb: fichero
     Untitled2.ipynb: fichero
     Untitled3.ipynb: fichero
     usa.ipynb: fichero


**Ejercicio 2** Escribir una función

*Nombre*: `descarga(path,url)`
- `path`: el path donde se alojará el fichero descargado, incluye el nombre local del fichero
- `url`: la url de un fichero

*Salida*: no devuelve nada, solo descarga usando requests el fichero indicado

Pista: en la solución de la práctica 4, ejercicio 2, podemos ver un ejemplo de descarga

In [8]:
import requests

# solución
def descarga(path,url):
    r = requests.get(url, allow_redirects=True) # el fichero queda en la variable r
    if r.status_code == 200:
        with open(path, 'wb') as f:
            f.write(r.content) # ahora lo grabamos localmente
    else:
        print("Error ",r.status_code)

# no tocar a partir de aquí, es para probar descargando dos ficheros, clientes.csv y gestores.csv que vamos a usar a continuación
url_clientes = "https://raw.githubusercontent.com/RafaelCaballero/tdm/master/datos/clientes2.csv"  
path_raw = Path(path,"raw")
path_clientes = Path(path_raw,"clientes.csv")
descarga(path_clientes,url_clientes)
path_gestores = Path(path_raw,"gestores.csv")
url_gestores = "https://raw.githubusercontent.com/RafaelCaballero/tdm/master/datos/clientegestor.csv"
descarga(path_gestores,url_gestores)
lista(path_raw)

d:\docencia\2223\APD\pruebas\raw
     abr_meteo22.csv: fichero
     ago_meteo22.csv: fichero
     clientes.csv: fichero
     dic_meteo22.csv: fichero
     download: carpeta
     ene_meteo22.csv: fichero
     feb_meteo22.csv: fichero
     gestores.csv: fichero
     Interpretación_datos_meteorologicos.pdf: fichero
     jul_meteo22.csv: fichero
     jun_meteo22.csv: fichero
     mar_meteo22.csv: fichero
     may_meteo22.csv: fichero
     nov_meteo22.csv: fichero
     oct_meteo22.csv: fichero
     readme.txt: fichero
     sep_meteo22.csv: fichero


**Ejercicio 3** Empezamos cargando los dos ficheros que acabamos de descargar:

In [9]:
df_clientes = pd.read_csv(path_clientes)
df_gestores = pd.read_csv(path_gestores)
display_side_by_side(df_clientes,df_gestores)

cliente 
 documento 
 entregado 
 tasas 
 
 
 
 
 0 
 2910167 
 1687488 
 0 
 464 
 
 
 1 
 5799708 
 152861 
 0 
 112 
 
 
 2 
 7248498 
 1789308 
 0 
 730 
 
 
 3 
 5799708 
 1082327 
 0 
 621 
 
 
 4 
 6989459 
 1537165 
 0 
 496 
 
 
 5 
 8474202 
 1938730 
 0 
 974 
 
 
 6 
 6945307 
 89214 
 1 
 466 
 
 
 7 
 3175818 
 1453012 
 0 
 298 
 
 
 8 
 7248498 
 349342 
 1 
 312 
 
 
 9 
 5651729 
 1952321 
 1 
 811 
 
 
 10 
 3515207 
 1667897 
 0 
 911 
 
 
 11 
 5799708 
 618022 
 0 
 426 
 
 
 12 
 6945307 
 731554 
 1 
 365 
 
 
 13 
 5799708 
 661854 
 1 
 936 
 
 
 14 
 3175818 
 1340745 
 1 
 148 
 
 
 15 
 4225453 
 538570 
 0 
 594 
 
 
 16 
 4225453 
 402928 
 0 
 360 
 
 
 17 
 4225453 
 1983821 
 0 
 849 
 
 
 18 
 3175818 
 248678 
 1 
 581 
 
 
 19 
 3515207 
 772334 
 0 
 730 
 
 
 20 
 2910167 
 684909 
 0 
 125 
 
 
 21 
 6989459 
 766933 
 0 
 413 
 
 
 22 
 7248498 
 1155206 
 0 
 224 
 
 
 23 
 5799708 
 301811 
 0 
 882 
 
 
 24 
 3515207 
 1544285 
 0 
 833 
 
 
 25 
 7248498 
 1502403 
 0 
 520 
 
 
 26 
 6945307 
 138418 
 0 
 845 
 
 
 27 
 5799708 
 1298575 
 0 
 889 
 
 
 28 
 2910167 
 796437 
 1 
 914 
 
 
 29 
 2910167 
 1268173 
 1 
 787 
 
 
 30 
 6989459 
 813491 
 0 
 767 
 
 
 31 
 6945307 
 1151582 
 0 
 547 
 
 
 32 
 2910167 
 1468405 
 1 
 222 
 
 
 33 
 4225453 
 374965 
 1 
 532 
 
 
 34 
 5799708 
 1119702 
 1 
 572 
 
 
 35 
 5651729 
 1184344 
 0 
 997 
 
 
 36 
 6989459 
 1374634 
 0 
 169 
 
 
 37 
 3175818 
 1658931 
 1 
 660 
 
 
 38 
 4225453 
 1886646 
 1 
 365 
 
 
 39 
 3175818 
 246532 
 1 
 995 
 
 
 40 
 6989459 
 1580475 
 0 
 108 
 
 
 41 
 3175818 
 306209 
 1 
 810 
 
 
 42 
 8474202 
 337032 
 0 
 268 
 
 
 43 
 7248498 
 1267571 
 0 
 238 
 
 
 44 
 3515207 
 1823510 
 0 
 176 
 
 
 45 
 5651729 
 865714 
 1 
 128 
 
 
 46 
 4225453 
 1174145 
 0 
 940 
 
 
 47 
 3175818 
 1973028 
 1 
 421 
 
 
 48 
 5799708 
 1736433 
 1 
 807 
 
 
 49 
 8474202 
 923338 
 0 
 522 
 
 
 50 
 5651729 
 1855090 
 0 
 505 
 
 
 51 
 2910167 
 610267 
 1 
 404 
 
 
 52 
 3175818 
 1018113 
 0 
 123 
 
 
 53 
 3175818 
 1099908 
 1 
 764 
 
 
 54 
 4225453 
 23700 
 1 
 489 
 
 
 55 
 5799708 
 1043315 
 0 
 249 
 
 
 56 
 4225453 
 606046 
 1 
 183 
 
 
 57 
 8474202 
 1502383 
 1 
 910 
 
 
 58 
 5651729 
 1222646 
 1 
 250 
 
 
 59 
 8474202 
 648137 
 0 
 917 
 
 
 60 
 4225453 
 146316 
 1 
 144 
 
 
 61 
 3515207 
 1617562 
 1 
 517 
 
 
 62 
 6945307 
 871912 
 0 
 733 
 
 
 63 
 3515207 
 1864397 
 1 
 394 
 
 
 64 
 5651729 
 660580 
 0 
 649 
 
 
 65 
 4225453 
 1326155 
 0 
 364 
 
 
 66 
 5651729 
 1832760 
 0 
 425 
 
 
 67 
 5799708 
 395151 
 0 
 104 
 
 
 68 
 5651729 
 390039 
 0 
 330 
 
 
 69 
 7248498 
 367156 
 0 
 910 
 
 
 70 
 6989459 
 203130 
 1 
 178 
 
 
 71 
 6945307 
 1096270 
 1 
 282 
 
 
 72 
 2910167 
 569935 
 0 
 817 
 
 
 73 
 8474202 
 631878 
 1 
 640 
 
 
 74 
 6989459 
 167138 
 0 
 685 
 
 
 75 
 3175818 
 175572 
 0 
 554 
 
 
 76 
 8474202 
 1840289 
 1 
 742 
 
 
 77 
 2910167 
 29738 
 0 
 113 
 
 
 78 
 6989459 
 1088640 
 0 
 743 
 
 
 79 
 3515207 
 516112 
 1 
 215 
 
 
 80 
 6945307 
 696285 
 1 
 647 
 
 
 81 
 4225453 
 1972874 
 1 
 458 
 
 
 82 
 8474202 
 1980829 
 1 
 405 
 
 
 83 
 5651729 
 1186057 
 0 
 322 
 
 
 84 
 7248498 
 1777385 
 0 
 826 
 
 
 85 
 2910167 
 776034 
 1 
 555 
 
 
 86 
 7248498 
 1083094 
 0 
 860 
 
 
 87 
 4225453 
 822770 
 0 
 715 
 
 
 88 
 5651729 
 1266901 
 0 
 987 
 
 
 89 
 4225453 
 131792 
 0 
 543 
 
 
 90 
 6989459 
 453257 
 1 
 187 
 
 
 91 
 6945307 
 502116 
 0 
 225 
 
 
 92 
 4225453 
 1102905 
 1 
 816 
 
 
 93 
 3515207 
 1422888 
 0 
 589 
 
 
 94 
 6945307 
 1397901 
 0 
 837 
 
 
 95 
 5651729 
 1627790 
 1 
 515 
 
 
 96 
 6989459 
 164481 
 1 
 942 
 
 
 97 
 5799708 
 1550749 
 0 
 774 
 
 
 98 
 3175818 
 871520 
 0 
 233 
 
 
 99 
 7248498 
 225806 
 0 
 237 
 
 
 100 
 5651729 
 84621 
 0 
 279 
 
 
 101 
 7248498 
 27282 
 1 
 584 
 
 
 102 
 5651729 
 1910255 
 0 
 127 


Escribir código para generar un nuevo dataframe `df_gc` que combine los dos dataframes añadiendo a los datos de `df_clientes` su gestor. Si un cliente no tiene gestor se eliminará del resultado (igualmente no aparecerán gestores que no sean de clientes en `df_clientes`)

In [10]:
df_gc = pd.merge(df_clientes,df_gestores)
# para probar
df_gc

cliente  documento  entregado  tasas gestor
0    2910167    1687488          0    464      A
1    2910167     684909          0    125      A
2    2910167     796437          1    914      A
3    2910167    1268173          1    787      A
4    2910167    1468405          1    222      A
..       ...        ...        ...    ...    ...
495  4225453     746041          1    718      A
496  4225453     389625          0    506      A
497  4225453     234715          0    785      A
498  4225453     770521          1    836      A
499  4225453    1835650          1    851      A

[500 rows x 5 columns]

In [11]:
df_gc.gestor.value_counts()

A    244
B    154
C    102
Name: gestor, dtype: int64

**Ejercicio 4** (2 puntos) Queremos repartir el dataframe df_gc en varios, uno por cada gestor, que además se grabarán en la carpeta `repartidos` con nombre del gestor (y extensión csv). Escribir una función 

*Nombre*: `reparte(path,df)`
- `path`: el path de una carpeta donde se alojarán los nuevos ficheros creados a partir de df, uno por gestor
- `df`: el dataframe a repartir en varios, debe incluir una columna `gestor`que se usara para repartir

*Salida*: no devuelve nada, solo descarga usando requests el fichero indicado

Idea: para cada valor diferente de la columna gestor (usar un `for`) se usara yun filtro para las filas del gestor. Se aplicará el filtro para obtener un dataframe nuevo con los datos del gestor, que a continuación se grabará en la carpeta indicada

In [12]:
# solución
def reparte(path,df):
    for g in df.gestor.unique():
        print(g)
        path_g = Path(path,g+".csv")
        filtro_g = df.gestor==g
        df_g  = df[filtro_g]
        df_g.to_csv(path_g,index=False)
        
## llamada         
path_repartidos = Path(path,"repartidos")
reparte(path_repartidos,df_gc)
# para comprobar
lista(path_repartidos)

A
B
C
d:\docencia\2223\APD\pruebas\repartidos
     A.csv: fichero
     B.csv: fichero
     C.csv: fichero


**Ejercicio 5** Recordamos el dataframe df_usa con opiniones de usuarios de Twitter sobre los candidatos de 2016

In [13]:

df_usa

id   clinton     trump
0       729822791747444737 -0.567568  0.415094
1               3438745348 -0.666667 -0.111111
2               1167485504 -0.466667  0.500000
3                219793074 -0.729730  0.620370
4               4605113535 -0.586207  0.373832
...                    ...       ...       ...
161168           118676558  0.000000  0.000000
161169          1438467932  0.857143  0.000000
161170  772542459901571072 -0.583333  0.000000
161171          3833168952 -0.400000  0.000000
161172  797601605449154560  0.017241  0.000000

[161173 rows x 3 columns]

En cada fila `clinton`, `trump` corresponden a la opinión media sobre los candidatos Clinton y Trump que tuvo el usuario de id `id`en la campaña para las elecciones USA 2016. Tanto clinton como trump van de -1 (opinión totalmente negativa) a +1 totalmente positiva.
Queremos crear un columna nueva C2 que valga -1 para las filas con clinton<0, +1 para las filas con clinton>0 y 0 si clinton==0. Una forma de hacerlo es crear la columna con un valor inicial y luego cambiar con filtros este valor inicial para que tome el que queremos

In [14]:
df_usa["C2"] = 0 # la inicializamos a 0 
df_usa

id   clinton     trump  C2
0       729822791747444737 -0.567568  0.415094   0
1               3438745348 -0.666667 -0.111111   0
2               1167485504 -0.466667  0.500000   0
3                219793074 -0.729730  0.620370   0
4               4605113535 -0.586207  0.373832   0
...                    ...       ...       ...  ..
161168           118676558  0.000000  0.000000   0
161169          1438467932  0.857143  0.000000   0
161170  772542459901571072 -0.583333  0.000000   0
161171          3833168952 -0.400000  0.000000   0
161172  797601605449154560  0.017241  0.000000   0

[161173 rows x 4 columns]

Completar el código para que C2 tome los valores que deseamos usando filtros

Pista: `df_usa.loc[filtro,"C2"]` accede a la columna "C2" de las filas de `df_usa` que verifican el filtro 

In [15]:
# solución
filtronegativo = df_usa.clinton<0
filtropositivo = df_usa.clinton>0

df_usa.loc[filtronegativo,"C2"] = -1
df_usa.loc[filtropositivo,"C2"] = +1

# para probar
df_usa



id   clinton     trump  C2
0       729822791747444737 -0.567568  0.415094  -1
1               3438745348 -0.666667 -0.111111  -1
2               1167485504 -0.466667  0.500000  -1
3                219793074 -0.729730  0.620370  -1
4               4605113535 -0.586207  0.373832  -1
...                    ...       ...       ...  ..
161168           118676558  0.000000  0.000000   0
161169          1438467932  0.857143  0.000000   1
161170  772542459901571072 -0.583333  0.000000  -1
161171          3833168952 -0.400000  0.000000  -1
161172  797601605449154560  0.017241  0.000000   1

[161173 rows x 4 columns]

### map

El método anterior para crear una columna nueva puede llegar a ser muy complicado; por ello se suele usar `map`. Vamos a verlo para hacer algo análogo con la columna T2

La idea es la siguiente: primero se escribe una función en Python que devuelva el valor nuevo de la columna T2 para un valor concreto de T (es la idea de siempre, pensar en números y luego extender a columnas/filas)

In [16]:
def posneg(t):
    if t<0:
        res=-1
    elif t==0:
        res=0
    else:
        res=+1
    return res

# para probar
print("-0.85 --> ",posneg(-0.85))
print("+0.15 --> ",posneg(+0.15))
print("+0 --> ",posneg(+0))

        

-0.85 -->  -1
+0.15 -->  1
+0 -->  0



Una vez que hemos visto que funciona ya podemos crear la nueva columna T2 que será el resultado de aplicar la función `posneg` a cada valor de la columna T, uno a uno. La función `map` de `Series`nos permite hacer eso

In [17]:
df_usa["T2"] = df_usa.trump.map(posneg)
df_usa

id   clinton     trump  C2  T2
0       729822791747444737 -0.567568  0.415094  -1   1
1               3438745348 -0.666667 -0.111111  -1  -1
2               1167485504 -0.466667  0.500000  -1   1
3                219793074 -0.729730  0.620370  -1   1
4               4605113535 -0.586207  0.373832  -1   1
...                    ...       ...       ...  ..  ..
161168           118676558  0.000000  0.000000   0   0
161169          1438467932  0.857143  0.000000   1   0
161170  772542459901571072 -0.583333  0.000000  -1   0
161171          3833168952 -0.400000  0.000000  -1   0
161172  797601605449154560  0.017241  0.000000   1   0

[161173 rows x 5 columns]

**Ejercicio 6** [2 puntos] Consideramos ahora el siguiente dataframe con datos de usuarios en Twitter

In [18]:
url = "https://raw.githubusercontent.com/RafaelCaballero/tdm/master/datos/user_coords.csv"
import pandas as pd
df_coord = pd.read_csv(url).drop(columns=["level_0","index"])
df_coord

_id  followers                        location  \
0             148076807        515            19.447297,-70.678351   
1             557076972       1022                             BCN   
2    800430827947442176         20                Valencia, España   
3            1187834312        653                   Madrid, Spain   
4            3153928585         99               Málaga, Andalucía   
5              18901057       1562                UK/Brasil/España   
6             117884436       1502                       Barcelona   
7    761237429072388096         35     Madrid, Comunidad de Madrid   
8             184959473        733            LA MANCHA y SEVILLA    
9            2838312285         27                          España   
10            984999942       1239                           Elche   
11            287826955       2244                       Salamanca   
12            292027758       2442                 madrid (españa)   
13            459477808        697               Estadio de Riazor   
14            875433438        491        De Salamanca, en Madrid    
15             89165765      27536                          Madrid   
16           2378742043        106                        valencia   
17   701379658600882176        168                   Elche, España   
18             42004114       2340           Canary Islands, Spain   
19            299264539        147  Santa Cruz de Tenerife, España   
20             96663968       1394               Budapest, Hungría   
21   913311027407802368         44                  Madrid, España   
22            180333341        413                Valencia, España   
23           1974659378       2198                       Barcelona   
24   745651197286162432        203                 Carmona, España   
25             92629354        981                Zaragoza, España   
26           4069778056        702                   Madrid, Spain   
27            996055734        475  Móstoles-Madrid/Orihuela Costa   
28            278039896        241                       Barcelona   
29            118352921       2961                   Madrid ,Spain   
30            270010775         39              Felgosas-A Coruña    
31            184537316        408                        PARAGUAY   
32           1580184631        102                          Elche.   
33   719905537396228096        706                    Mijas, Spain   
34           1495328876        763     Madrid, Comunidad de Madrid   
35            246837853       2480                          Madrid   
36            142451580         52                          Madrid   
37            709714971        872                  Málaga, España   
38             15703541       5084                Washington, D.C.   
39           1553476147        214                  Madrid - Spain   
40   849681349162283008        217  Principado de Asturias, España   
41            757471410        324               Glasgow, Scotland   
42           1122025458        846                A Coruña, España   
43            233345061       2437                        Albacete   
44           1723462104        103               Torrent, Valencia   
45             84143687       1814                Madrid - Cáceres   
46            106841104         90             Cuidotumascota.com    
47            943230391         64                         Arteixo   
48  1337479172441169920       1328                 Sevilla, España   
49            446433608        924                   Villalba Sity   
50            490529286        431                         JLAM  ∞   
51           3352643962        256                Valencia, España   
52            799512942       1149      Las Palmas, Islas Canarias   

        screen_name  verified       long        lat  
0       eugenio_sky     False  -3.685060  40.427170  
1       annetapedro     False   2.176402  41.405603  
2     Trastarteando     False  -0.293131  39.567629  
3        manukareol    

Añadir a `df_coord` nueva columna `muchos_followers` que valga "Sí" si el número de followers es mayor 1000 o "No" en otro caso. Hacerlo en dos pasos.

a) Escribir una función de Pythom `muchos` que reciba como parámetro un valor n y devuelva "Sí" si n>1000, y "No" en otro caso

In [19]:
def muchos(n): #completar
    if n>1000:
        res="Sí"
    else:
        res="No"
    return res
    
# para probar  
print("muchos(2000)",muchos(2000)) # Sí
print("muchos(500)",muchos(500)) # No
print("muchos(1000)",muchos(1000)) # No

muchos(2000) Sí
muchos(500) No
muchos(1000) No


b) Aplicar con `map`la función a todos los elementos de `followers` para crear la columna `muchos_followers` en `df_coord`

In [20]:
# solución
df_coord["muchos_followers"] = df_coord.followers.map(muchos)

# para comprobar
df_coord

_id  followers                        location  \
0             148076807        515            19.447297,-70.678351   
1             557076972       1022                             BCN   
2    800430827947442176         20                Valencia, España   
3            1187834312        653                   Madrid, Spain   
4            3153928585         99               Málaga, Andalucía   
5              18901057       1562                UK/Brasil/España   
6             117884436       1502                       Barcelona   
7    761237429072388096         35     Madrid, Comunidad de Madrid   
8             184959473        733            LA MANCHA y SEVILLA    
9            2838312285         27                          España   
10            984999942       1239                           Elche   
11            287826955       2244                       Salamanca   
12            292027758       2442                 madrid (españa)   
13            459477808        697               Estadio de Riazor   
14            875433438        491        De Salamanca, en Madrid    
15             89165765      27536                          Madrid   
16           2378742043        106                        valencia   
17   701379658600882176        168                   Elche, España   
18             42004114       2340           Canary Islands, Spain   
19            299264539        147  Santa Cruz de Tenerife, España   
20             96663968       1394               Budapest, Hungría   
21   913311027407802368         44                  Madrid, España   
22            180333341        413                Valencia, España   
23           1974659378       2198                       Barcelona   
24   745651197286162432        203                 Carmona, España   
25             92629354        981                Zaragoza, España   
26           4069778056        702                   Madrid, Spain   
27            996055734        475  Móstoles-Madrid/Orihuela Costa   
28            278039896        241                       Barcelona   
29            118352921       2961                   Madrid ,Spain   
30            270010775         39              Felgosas-A Coruña    
31            184537316        408                        PARAGUAY   
32           1580184631        102                          Elche.   
33   719905537396228096        706                    Mijas, Spain   
34           1495328876        763     Madrid, Comunidad de Madrid   
35            246837853       2480                          Madrid   
36            142451580         52                          Madrid   
37            709714971        872                  Málaga, España   
38             15703541       5084                Washington, D.C.   
39           1553476147        214                  Madrid - Spain   
40   849681349162283008        217  Principado de Asturias, España   
41            757471410        324               Glasgow, Scotland   
42           1122025458        846                A Coruña, España   
43            233345061       2437                        Albacete   
44           1723462104        103               Torrent, Valencia   
45             84143687       1814                Madrid - Cáceres   
46            106841104         90             Cuidotumascota.com    
47            943230391         64                         Arteixo   
48  1337479172441169920       1328                 Sevilla, España   
49            446433608        924                   Villalba Sity   
50            490529286        431                         JLAM  ∞   
51           3352643962        256                Valencia, España   
52            799512942       1149      Las Palmas, Islas Canarias   

        screen_name  verified       long        lat muchos_followers  
0       eugenio_sky     False  -3.685060  40.427170               No  
1       annetapedro     False   2.176402  41.405603               Sí  
2     Trastarteando     Fa

### apply 
`map` es muy útil, pero a veces resulta insuficiente. Supongamos que ahora en df_usa queremos añadir una columna "voto" que indique si el usuario es "Pro-trump" (trump>clinton), "pro-clinton" (clinton<trump) o "neutro". El problema es que ahora no podemos usar map porque se aplica a una columna y nuestra nueva columna depende de 2, trump y clinton. 

Por fortuna `apply` nos permite aplicar una función a una fila entera.

In [21]:
df_usa

id   clinton     trump  C2  T2
0       729822791747444737 -0.567568  0.415094  -1   1
1               3438745348 -0.666667 -0.111111  -1  -1
2               1167485504 -0.466667  0.500000  -1   1
3                219793074 -0.729730  0.620370  -1   1
4               4605113535 -0.586207  0.373832  -1   1
...                    ...       ...       ...  ..  ..
161168           118676558  0.000000  0.000000   0   0
161169          1438467932  0.857143  0.000000   1   0
161170  772542459901571072 -0.583333  0.000000  -1   0
161171          3833168952 -0.400000  0.000000  -1   0
161172  797601605449154560  0.017241  0.000000   1   0

[161173 rows x 5 columns]

In [22]:
def pro(row): 
    #print(row)
    if row["trump"]>row["clinton"]:
        res="Pro-trump"
    elif row["trump"]<row["clinton"]:
        res="Pro-clinton"
    else:
        res="neutro"
    return(res)
        
pro(df_usa.iloc[0])        

'Pro-trump'

Aplicamos la función

In [23]:
df_usa["voto"] = df_usa.apply(pro,axis=1) # obsérvese que ahora no se aplica a una columna sino al dataframe entero
df_usa

id   clinton     trump  C2  T2         voto
0       729822791747444737 -0.567568  0.415094  -1   1    Pro-trump
1               3438745348 -0.666667 -0.111111  -1  -1    Pro-trump
2               1167485504 -0.466667  0.500000  -1   1    Pro-trump
3                219793074 -0.729730  0.620370  -1   1    Pro-trump
4               4605113535 -0.586207  0.373832  -1   1    Pro-trump
...                    ...       ...       ...  ..  ..          ...
161168           118676558  0.000000  0.000000   0   0       neutro
161169          1438467932  0.857143  0.000000   1   0  Pro-clinton
161170  772542459901571072 -0.583333  0.000000  -1   0    Pro-trump
161171          3833168952 -0.400000  0.000000  -1   0    Pro-trump
161172  797601605449154560  0.017241  0.000000   1   0  Pro-clinton

[161173 rows x 6 columns]

**Ejercicio 7** (2 puntos) En `df_coord` añadir una columna `coords` que contendrá un string de la forma "lat, long" para cada pareja de coordenadas

In [24]:
df_coord

_id  followers                        location  \
0             148076807        515            19.447297,-70.678351   
1             557076972       1022                             BCN   
2    800430827947442176         20                Valencia, España   
3            1187834312        653                   Madrid, Spain   
4            3153928585         99               Málaga, Andalucía   
5              18901057       1562                UK/Brasil/España   
6             117884436       1502                       Barcelona   
7    761237429072388096         35     Madrid, Comunidad de Madrid   
8             184959473        733            LA MANCHA y SEVILLA    
9            2838312285         27                          España   
10            984999942       1239                           Elche   
11            287826955       2244                       Salamanca   
12            292027758       2442                 madrid (españa)   
13            459477808        697               Estadio de Riazor   
14            875433438        491        De Salamanca, en Madrid    
15             89165765      27536                          Madrid   
16           2378742043        106                        valencia   
17   701379658600882176        168                   Elche, España   
18             42004114       2340           Canary Islands, Spain   
19            299264539        147  Santa Cruz de Tenerife, España   
20             96663968       1394               Budapest, Hungría   
21   913311027407802368         44                  Madrid, España   
22            180333341        413                Valencia, España   
23           1974659378       2198                       Barcelona   
24   745651197286162432        203                 Carmona, España   
25             92629354        981                Zaragoza, España   
26           4069778056        702                   Madrid, Spain   
27            996055734        475  Móstoles-Madrid/Orihuela Costa   
28            278039896        241                       Barcelona   
29            118352921       2961                   Madrid ,Spain   
30            270010775         39              Felgosas-A Coruña    
31            184537316        408                        PARAGUAY   
32           1580184631        102                          Elche.   
33   719905537396228096        706                    Mijas, Spain   
34           1495328876        763     Madrid, Comunidad de Madrid   
35            246837853       2480                          Madrid   
36            142451580         52                          Madrid   
37            709714971        872                  Málaga, España   
38             15703541       5084                Washington, D.C.   
39           1553476147        214                  Madrid - Spain   
40   849681349162283008        217  Principado de Asturias, España   
41            757471410        324               Glasgow, Scotland   
42           1122025458        846                A Coruña, España   
43            233345061       2437                        Albacete   
44           1723462104        103               Torrent, Valencia   
45             84143687       1814                Madrid - Cáceres   
46            106841104         90             Cuidotumascota.com    
47            943230391         64                         Arteixo   
48  1337479172441169920       1328                 Sevilla, España   
49            446433608        924                   Villalba Sity   
50            490529286        431                         JLAM  ∞   
51           3352643962        256                Valencia, España   
52            799512942       1149      Las Palmas, Islas Canarias   

        screen_name  verified       long        lat muchos_followers  
0       eugenio_sky     False  -3.685060  40.427170               No  
1       annetapedro     False   2.176402  41.405603               Sí  
2     Trastarteando     Fa

En la primera fila, por ejemplo, la columna valdrá "40.427170 , -3.685060". Utilizar `apply`y una función auxiliar
`prepara_coords`.

Pista: recordar que str(x) convirte el número x e un string

In [25]:
def prepara_coords(row): #completar
    return str(row["lat"])+", "+str(row["long"])
df_coord["coords"] = df_coord.apply(prepara_coords,axis=1)

# para comprobar
df_coord

_id  followers                        location  \
0             148076807        515            19.447297,-70.678351   
1             557076972       1022                             BCN   
2    800430827947442176         20                Valencia, España   
3            1187834312        653                   Madrid, Spain   
4            3153928585         99               Málaga, Andalucía   
5              18901057       1562                UK/Brasil/España   
6             117884436       1502                       Barcelona   
7    761237429072388096         35     Madrid, Comunidad de Madrid   
8             184959473        733            LA MANCHA y SEVILLA    
9            2838312285         27                          España   
10            984999942       1239                           Elche   
11            287826955       2244                       Salamanca   
12            292027758       2442                 madrid (españa)   
13            459477808        697               Estadio de Riazor   
14            875433438        491        De Salamanca, en Madrid    
15             89165765      27536                          Madrid   
16           2378742043        106                        valencia   
17   701379658600882176        168                   Elche, España   
18             42004114       2340           Canary Islands, Spain   
19            299264539        147  Santa Cruz de Tenerife, España   
20             96663968       1394               Budapest, Hungría   
21   913311027407802368         44                  Madrid, España   
22            180333341        413                Valencia, España   
23           1974659378       2198                       Barcelona   
24   745651197286162432        203                 Carmona, España   
25             92629354        981                Zaragoza, España   
26           4069778056        702                   Madrid, Spain   
27            996055734        475  Móstoles-Madrid/Orihuela Costa   
28            278039896        241                       Barcelona   
29            118352921       2961                   Madrid ,Spain   
30            270010775         39              Felgosas-A Coruña    
31            184537316        408                        PARAGUAY   
32           1580184631        102                          Elche.   
33   719905537396228096        706                    Mijas, Spain   
34           1495328876        763     Madrid, Comunidad de Madrid   
35            246837853       2480                          Madrid   
36            142451580         52                          Madrid   
37            709714971        872                  Málaga, España   
38             15703541       5084                Washington, D.C.   
39           1553476147        214                  Madrid - Spain   
40   849681349162283008        217  Principado de Asturias, España   
41            757471410        324               Glasgow, Scotland   
42           1122025458        846                A Coruña, España   
43            233345061       2437                        Albacete   
44           1723462104        103               Torrent, Valencia   
45             84143687       1814                Madrid - Cáceres   
46            106841104         90             Cuidotumascota.com    
47            943230391         64                         Arteixo   
48  1337479172441169920       1328                 Sevilla, España   
49            446433608        924                   Villalba Sity   
50            490529286        431                         JLAM  ∞   
51           3352643962        256                Valencia, España   
52            799512942       1149      Las Palmas, Islas Canarias   

        screen_name  verified       long        lat muchos_followers  \
0       eugenio_sky     False  -3.685060  40.427170               No   
1       annetapedro     False   2.176402  41.405603               Sí   
2     Trastarteando    

Una curiosidad final: vamos a mostrar  dónde está cada usuario (si esto no funciona no pasa nada es solo ilustrativo)

In [28]:
import folium

m = folium.Map(zoom_start=5,location=[37,-10])
for i,fila in df_coord.iterrows():
    # añadir marcador al mapa
    marcador = folium.Marker(location=[fila["lat"], fila["long"]], popup=fila["screen_name"] )
    marcador.add_to(m)
   
# mostramos el mapa
m